File: data/readable_format/fortran/file343.f90
PROGRAM test_zlaswpf
  IMPLICIT NONE
  INTEGER, PARAMETER :: n = 4
  INTEGER :: ipiv(4) = (/ 2, 3, 4, 1 /)
  COMPLEX*16 :: A(4,4)
  INTEGER :: i, j

  ! Initialize the matrix A
  DO i = 1, n
     DO j = 1, n
        A(i,j) = COMPLEX(i, j)
     END DO
  END DO

  ! Call the subroutine
  CALL ZLASWPF(n, A, 4, 1, 4, ipiv, 1)

  ! Print the result
  PRINT *, 'Resulting Matrix A:'
  DO i = 1, n
     PRINT '(4(F6.1,A,F6.1,A,3X))', (REAL(A(i,j)), ' + i', AIMAG(A(i,j)), ',', j=1,n)
  END DO

END PROGRAM test_zlaswpf

SUBROUTINE ZLASWPF( N, A, LDA, K1, K2, IPIV, INCX )
  INTEGER            INCX, K1, K2, LDA, N
  INTEGER            IPIV( * )
  COMPLEX*16         A( LDA, * )
  INTEGER            I, I1, I2, INC, IP, IX, IX0, J, K, N32
  COMPLEX*16         TEMP

  IF( INCX.GT.0 ) THEN
     IX0 = K1
     I1 = K1
     I2 = K2
     INC = 1
  ELSE IF( INCX.LT.0 ) THEN
     IX0 = 1 + ( 1-K2 )*INCX
     I1 = K2
     I2 = K1
     INC = -1
  ELSE
     RETURN
  END IF

  N32 = ( N / 32 )*32
  IF( N32.NE.0 ) THEN
     DO 30 J = 1, N32, 32
        IX = IX0
        DO 20 I = I1, I2, INC
           IP = IPIV( IX )
           IF( IP.NE.I ) THEN
              DO 10 K = J, J + 31
                 TEMP = A( I, K )
                 A( I, K ) = A( IP, K )
                 A( IP, K ) = TEMP
10             CONTINUE
           END IF
           IX = IX + INCX
20       CONTINUE
30    CONTINUE
  END IF
  IF( N32.NE.N ) THEN
     N32 = N32 + 1
     IX = IX0
     DO 50 I = I1, I2, INC
        IP = IPIV( IX )
        IF( IP.NE.I ) THEN
           DO 40 K = N32, N
              TEMP = A( I, K )
              A( I, K ) = A( IP, K )
              A( IP, K ) = TEMP
40          CONTINUE
        END IF
        IX = IX + INCX
50    CONTINUE
  END IF

  RETURN
END SUBROUTINE ZLASWPF

File: data/readable_format/cpp/file343.cpp
#include <iostream>
#include <vector>
#include <complex>

// Forward declaration of ZLASWPF
void ZLASWPF(int n, std::vector<std::vector<std::complex<double>>>& A, int lda, int k1, int k2, const std::vector<int>& ipiv, int incx);

int main() {
    const int n = 4; // Matrix size
    std::vector<int> ipiv = {2, 3, 4, 1}; // Permutation vector
    std::vector<std::vector<std::complex<double>>> A(n, std::vector<std::complex<double>>(n));

    // Initialize the matrix A
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            A[i][j] = std::complex<double>(i+1, j+1);
        }
    }

    // Call the ZLASWPF function
    ZLASWPF(n, A, n, 1, 4, ipiv, 1);

    // Print the resulting matrix
    std::cout << "Resulting Matrix A:" << std::endl;
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            std::cout << A[i][j] << "\t";
        }
        std::cout << std::endl;
    }

    return 0;
}

// Implementation of ZLASWPF
void ZLASWPF(int n, std::vector<std::vector<std::complex<double>>>& A, int lda, int k1, int k2, const std::vector<int>& ipiv, int incx) {
    int i, i1, i2, inc, ip, ix, ix0, j, k, n32;
    std::complex<double> temp;

    if (incx > 0) {
        ix0 = k1;
        i1 = k1;
        i2 = k2;
        inc = 1;
    } else if (incx < 0) {
        ix0 = 1 + (1 - k2) * incx;
        i1 = k2;
        i2 = k1;
        inc = -1;
    } else {
        return; // Do nothing if incx is zero
    }

    n32 = (n / 32) * 32;
    if (n32 != 0) {
        for (j = 0; j < n32; j += 32) {
            ix = ix0;
            for (i = i1; i <= i2; i += inc) {
                ip = ipiv[ix - 1] - 1; // Adjust for 0-based indexing
                if (ip != i - 1) {
                    for (k = j; k < j + 32; ++k) {
                        temp = A[i - 1][k];
                        A[i - 1][k] = A[ip][k];
                        A[ip][k] = temp;
                    }
                }
                ix += incx;
            }
        }
    }

    if (n32 != n) {
        n32 = n32 + 1;
        ix = ix0;
        for (i = i1; i <= i2; i += inc) {
            ip = ipiv[ix - 1] - 1; // Adjust for 0-based indexing
            if (ip != i - 1) {
                for (k = n32 - 1; k < n; ++k) {
                    temp = A[i - 1][k];
                    A[i - 1][k] = A[ip][k];
                    A[ip][k] = temp;
                }
            }
            ix += incx;
        }
    }
}